internal/temporalhistoryv1/load.go

```diff
@@ -0,0 +1,162 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// !!! This file is copied from internal/temporalhistoryv1 to history/v1.
+// !!! DO NOT EDIT at history/v1/load.go.
+
+package history
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/iancoleman/strcase"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+const pathSep = "."
+
+type enumFix struct {
+	Path       string
+	TypePrefix string
+}
+
+// Map of JSON path to SCREAMING_CASE type prefix for proto enums
+// Not initialized until prepareFixes.Do(...) has been called
+var historyPathsToFix = make(map[string]string)
+var prepareFixes sync.Once
+
+// Recursively walk the protoreflect descriptions for the History object
+// collecting all enum paths for later fixing
+func collectPaths(path []string, fd protoreflect.FieldDescriptor) {
+	if fe := fd.Enum(); fe != nil {
+		for i := 0; i < fe.Values().Len(); i++ {
+			fullPath := strings.Join(append(path, fd.JSONName()), pathSep)
+			historyPathsToFix[fullPath] = strcase.ToScreamingSnake(string(fe.Name()))
+		}
+	}
+
+	if fm := fd.Message(); fm != nil {
+		path = append(path, fd.JSONName())
+		for i := 0; i < fm.Fields().Len(); i++ {
+			cd := fm.Fields().Get(i)
+			// Avoid blowing the stack when a proto may contain itself
+			// I'm looking at you temporal.api.failure.v1.Failure...
+			if cd.Name() == fd.Name() {
+				continue
+			}
+			collectPaths(path, fm.Fields().Get(i))
+		}
+
+	}
+}
+
+func visitAndFix(val interface{}, path string, fixes map[string]string) {
+	switch v := val.(type) {
+	case []interface{}:
+		for i := range v {
+			if vv, ok := v[i].(map[string]interface{}); ok {
+				visitAndFix(vv, path, fixes)
+			}
+		}
+	case map[string]interface{}:
+		for kk, vv := range v {
+			fullPath := fmt.Sprintf("%s.%s", path, kk)
+			switch vvv := vv.(type) {
+			case string:
+				if typePrefix, ok := fixes[fullPath]; ok {
+					newVal := strcase.ToScreamingSnake(vvv)
+					if !strings.HasPrefix(newVal, typePrefix) {
+						newVal = fmt.Sprintf("%s_%s", typePrefix, newVal)
+					}
+					v[kk] = newVal
+				}
+			case []interface{}:
+				visitAndFix(vvv, fullPath, fixes)
+			case map[string]interface{}:
+				visitAndFix(vvv, fullPath, fixes)
+			}
+		}
+	}
+}
+
+// LoadFromJSON deserializes history from a reader of JSON bytes. This does
+// not close the reader if it is closeable. This function is compatible both
+// with the "correct" SCREAMING_SNAKE enums of protojson as well as the old
+// PascalCase enums of Temporal's older history exports.
+func LoadFromJSON(r io.Reader, lastEventID int64) (*History, error) {
```

Comment:
I'd remove the `lastEventID` thing. That's a Go SDK only thing. We can change that SDK to lop events off after unmarshaling over there.

Reply:
Done!